virtual-display-driver-cli/src/main.rs

```diff
@@ -0,0 +1,343 @@
+        }
+        Command::AddMode {
+            id,
+            width,
+            height,
+            refresh_rate,
+        } => {
+            let mut client = Client::connect()?;
+            let mut monitor = client.get(id)?;
+
+            let new_mode_index = monitor.modes.len();
+            let new_mode = driver_ipc::Mode {
+                width,
+                height,
+                refresh_rates: refresh_rate,
+            };
+            monitor.modes.push(new_mode);
+            client.notify(vec![monitor])?;
+
+            if args.json {
+                let mut stdout = std::io::stdout().lock();
+                serde_json::to_writer_pretty(&mut stdout, &new_mode_index)?;
+            } else {
+                println!(
+                    "Added new mode {} to virtual monitor with ID {}.",
+                    new_mode_index.blue(),
+                    id.green()
+                );
+            }
+        }
+        Command::RemoveMode { id, mode_index } => {
+            let mut client = Client::connect()?;
+            let mut monitor = client.get(id)?;
+
+            if mode_index >= monitor.modes.len() {
+                eyre::bail!(
+                    "virtual monitor with ID {} has no mode with index {}",
+                    id,
+                    mode_index
+                );
+            }
+            if monitor.modes.len() == 1 {
+                eyre::bail!(
+                    "cannot remove last mode from virtual monitor with ID {}",
+                    id
+                );
+            }
+
+            monitor.modes.remove(mode_index);
+            client.notify(vec![monitor])?;
+
+            if args.json {
+                let mut stdout = std::io::stdout().lock();
+                serde_json::to_writer_pretty(&mut stdout, &mode_index)?;
+            } else {
+                println!(
+                    "Removed mode {} from virtual monitor with ID {}.",
+                    mode_index.blue(),
+                    id.green()
+                );
+            }
+        }
+        Command::Enable { id } => {
+            let mut client = Client::connect()?;
+            let outcome = set_enabled(&mut client, id, true)?;
+
+            if args.json {
+                let mut stdout = std::io::stdout().lock();
+                serde_json::to_writer_pretty(&mut stdout, &outcome)?;
+            } else {
+                let footnote = if outcome.toggled {
+                    ""
+                } else {
+                    " (was already enabled)"
+                };
+                println!("Enabled virtual monitor with ID {}{footnote}.", id.green());
+            }
+        }
+        Command::Disable { id } => {
+            let mut client = Client::connect()?;
+            let outcome = set_enabled(&mut client, id, false)?;
+
+            if args.json {
+                let mut stdout = std::io::stdout().lock();
+                serde_json::to_writer_pretty(&mut stdout, &outcome)?;
```

Comment:
I personally feel that pretty-printed output would be a better default (and users that want condensed output can pipe the output to `jq -c`), but I think the "right" option would just be to copy what other tools do. So, to address this, I'd want to just find some tools that have a JSON output option and figure out what format they use.

As for other output and errors, I think the important thing is that all output _on stdout_ is JSON-formatted. Tools can just check the status code is non-zero, then parse the output after-the-fact. I'd also lean towards errors staying as a human-readable output (which doesn't conflict with parsing since they'd be printed on stderr). I do feel confident saying that this is the more common option: plenty of tools always write to stdout with structured output (if enabled), and always write to stderr with human-readable output.

Reply:
> I personally feel that pretty-printed output would be a better default (and users that want condensed output can pipe the output to jq -c), but I think the "right" option would just be to copy what other tools do. So, to address this, I'd want to just find some tools that have a JSON output option and figure out what format they use.

That's fair! I don't have a super strong opinion either way. It is convenient that the output is easy to read (and parse in someone's head) when it's the default.

> I do feel confident saying that this is the more common option: plenty of tools always write to stdout with structured output (if enabled), and always write to stderr with human-readable output.

If that's the standard way of doing things, that's fine with me 😄